### DISCUSSION TITLE
Ignored files section is very slow (in git repo with GIT_DIR/GIT_WORKTREE)

### DISCUSSION THREAD

**Discussion by indigoviolet** [magit/magit#4750] [Support] [answered]
See #460 for context on the situation - I'm using a dotfiles manager (yadm) that sets up its git repo in one directory and uses $HOME as the work tree. Magit doesn't adequately support this situation, but that discussion is already in #460.

I use magit-status through a tramp session that intercepts calls to git and handles things correctly: 

```elisp
  (add-to-list 'tramp-methods
               '("yadm"
                 (tramp-login-program "yadm")
                 (tramp-login-args (("enter")))
                 (tramp-remote-shell "/bin/bash")
                 (tramp-remote-shell-args ("-c"))))
  (defun yadm-status ()
    (interactive)
    (require 'tramp)
    (with-current-buffer (magit-status "/yadm::")))
```

Anyway, my ignored files section is _very_ slow. I have tried to use `.dir-locals.el` to turn it off (following https://magit.vc/manual/magit/Per_002dRepository-Configuration.html), but I can't seem to get magit to see that override - I put it in $HOME/.dir-locals.el, but also tried putting it in the GIT_DIR. Neither seems to work.

How can I fix this?

Alternatively (additionally) - it would be nice to tell magit -- if this section gets longer than X, turn it off.

**Comment by tarsius** [comment, marked as answer]
> Neither seems to work.

Note sure, but this sounds like you might followed the instructions for `.gitconfig` when deciding where to place `.dir-locals.el`. That file should go to `/path/to/repo/.dir-locals.el`. First try to get it working on your local machine, if it works there but not over tramp, then it likely is a tramp issue (and tramp's documentation might be of help). I understand this involves symlinks, maybe Emacs doesn't follow the relevant links for some reason, and maybe that can be configured. Try with some simple local setting first, e.g., enable `hl-line-mode`.

> Alternatively (additionally) - it would be nice to tell magit -- if this section gets longer than X, turn it off.

I don't think that's currently feasible. We would end up switching from taking too long to display that section to taking just as long to decide that it takes too long to display and then *not even* displaying the costly result.

**Reply by indigoviolet** [reply]
Thanks, that was helpful advice. It turned out to be tramp's `enable-remote-dir-locals` variable, which needed to be `t`

**Comment by Thaodan** [comment]
Did you get it working?

Is this how your dir-locals.el looks like? 

Another thing I noticed that having `magit-add-section-hook` to insert `magit-insert-ignored-files` or `magit-insert-untracked-files` can be an issue here.

```
((magit-status-mode
  . ((eval . (magit-disable-section-inserter 'magit-insert-untracked-files)))))
```

**Comment by Thaodan** [comment]
Is it possible to do something like this without dir-locals? The issue is that dir locals also apply to all subdirectories.

EDIT: This one works for me:
```emacs-lisp
;; Small helper to launch Magit with yadm
(defun yadm ()
  (interactive)
  (let ((magit-status-sections-hook magit-status-sections-hook))
    (remove-hook 'magit-status-sections-hook 'magit-insert-untracked-files)
    (remove-hook 'magit-status-sections-hook 'magit-insert-ignored-files)
    (magit-status-setup-buffer "/yadm::")
    (setq magit-disabled-section-inserters '(magit-insert-untracked-files
                                                                            magit-insert-ignored-files))))

```

EDIT: Using dir-locals but using a directory class actually works much better similarly as the Magit manual states:
```emacs-lisp
     ;; Setup directory class for yadm tramp method so we don't so down magit with all the untracked and ignore files
     (dir-locals-set-class-variables 'magit-yadm
                                     '((magit-status-mode .
                                                          ((magit-disabled-section-inserters . (magit-insert-untracked-files
                                                                                                magit-insert-ignored-files))))))
     (dir-locals-set-directory-class "/yadm::" 'magit-yadm)
     ;; Add /yadm:: to directories that are always considered as safe ([[help:safe-local-variable-directories]])
     (setq safe-local-variable-directories (cl-union
                                            safe-local-variable-directories
                                            '("/yadm::")))
```
